Silence ResizeObservers while backgrounding YouTube on iOS.

sammygill opened the pull request:
#### e54ba88232e1805a9027fa848a4b3f8db79aa42a
<pre>
Silence ResizeObservers while backgrounding YouTube on iOS.
rdar://109837319
<a href="https://bugs.webkit.org/show_bug.cgi?id=258588">https://bugs.webkit.org/show_bug.cgi?id=258588</a>

Reviewed by Brent Fulgham.

Currently, if you background and foreground Safari on iPad while on
YouTube, you will get a different page layout compared to what was
displayed before the backgrounding process. Since the viewport sizes are
the same one would expect there to be no difference in the page layout.
This is due to a ResizeObserver callback using requestAnimationFrame to
execute some logic to reinitalize the ResizeObserver, but the observer
is ultimately put into a bad state due to the quick viewport size
changed. By silencing ResizeObservers during this process we should not
trigger this behavior and reintroduce the expected page layout again.

* Source/WebCore/dom/Document.cpp:
(WebCore::Document::updateResizeObservations):
* Source/WebCore/page/Quirks.cpp:
(WebCore::Quirks::shouldSilenceResizeObservers const):
* Source/WebCore/page/Quirks.h:

Canonical link: <a href="https://commits.webkit.org/265603@main">https://commits.webkit.org/265603@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8f4de017254592d5f5ad6fe7ef29660a92636982

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/11360 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/11570 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/11911 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/13008 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/10823 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/11381 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/13950 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/11553 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/13741 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/11522 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/12399 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/9617 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/13428 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/9694 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/10297 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/17493 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/10763 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/10454 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/13677 "11 api tests failed or timed out") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/10871 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/8950 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/10043 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/2727 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/14317 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/10726 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

brentfulgham commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -8602,6 +8602,10 @@ void Document::setHasSkippedResizeObservations(bool skipped)
 void Document::updateResizeObservations(Page& page)
 {
+    if (quirks().shouldSilenceResizeObservers()) {
+        addConsoleMessage(MessageSource::Other, MessageLevel::Info, "ResizeObservers silenced due to: rdar://110097836"_s);
```
This *should* only generate a message when the background occurs, so I don't think this will create too much logging.

smfr replied:
If we're going to log the console, we shouldn't log a radar number that is opaque to non-Apple folks. File a bugzilla, and reference that (if anything).

karlcow replied:
Agreed with Simon, we can create a public bug for it. 
Thanks @sgill26 for the log message ❤️ 
Do you want me to create the public bug?

karlcow replied:
see https://github.com/WebKit/WebKit/pull/15346#pullrequestreview-1502110638

sammygill replied:
Changed

brentfulgham reviewed (approved):
Looks good. r=me

smfr commented on Source/WebCore/page/Quirks.cpp:
```diff
@@ -718,6 +718,27 @@ bool Quirks::shouldOmitHTMLDocumentSupportedPropertyNames()
+        return false;
+
+    auto host = m_document->topDocument().url().host();
+    return equalLettersIgnoringASCIICase(host, "youtube.com"_s) || host.endsWithIgnoringASCIICase(".youtube.com"_s);
```
Elsewhere we do
```
    auto host = topDocument.url().host();
    auto isYouTube = host.endsWith(".youtube.com"_s) || host == "youtube.com"_s;
```
Why is this different?

sammygill replied:
I was trying to make it similar to the one we have for the Window resize events. We seem to do a few different things including `RegistrableDomain(m_document->url()).string()`

I'll double check which option is the correct/preferred way to do this host testing

karlcow replied:
I filed https://bugs.webkit.org/show_bug.cgi?id=258603

sammygill replied:
So it seems like checking the document url's host and using `RegistrableDomain` would both be valid for this scenario. I changed to using `RegistrableDomain` just to make it shorter

karlcow commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -8602,6 +8602,10 @@ void Document::setHasSkippedResizeObservations(bool skipped)
 void Document::updateResizeObservations(Page& page)
 {
+    if (quirks().shouldSilenceResizeObservers()) {
+        addConsoleMessage(MessageSource::Other, MessageLevel::Info, "ResizeObservers silenced due to: rdar://110097836"_s);
```
```suggestion
        addConsoleMessage(MessageSource::Other, MessageLevel::Info, "ResizeObservers silenced due to: http://webkit.org/b/258597"_s);
```

alanbaradlay replied:
or http://webkit.org/b/258597

karlcow reviewed (changes requested):
Opened the bug.

sammygill committed:
Silence ResizeObservers while backgrounding YouTube on iOS.
rdar://109837319
https://bugs.webkit.org/show_bug.cgi?id=258588

Reviewed by Brent Fulgham.

Currently, if you background and foreground Safari on iPad while on
YouTube, you will get a different page layout compared to what was
displayed before the backgrounding process. Since the viewport sizes are
the same one would expect there to be no difference in the page layout.
This is due to a ResizeObserver callback using requestAnimationFrame to
execute some logic to reinitalize the ResizeObserver, but the observer
is ultimately put into a bad state due to the quick viewport size
changed. By silencing ResizeObservers during this process we should not
trigger this behavior and reintroduce the expected page layout again.

* Source/WebCore/dom/Document.cpp:
(WebCore::Document::updateResizeObservations):
* Source/WebCore/page/Quirks.cpp:
(WebCore::Quirks::shouldSilenceResizeObservers const):
* Source/WebCore/page/Quirks.h:

Canonical link: https://commits.webkit.org/265603@main
